`Created by Ayushi Dubey`

# Importing Necessary Libraries

All the necessary libraries used in the notebook are imported in the below cell. The functions - `standardise_smiles` and `standardise_inchikey` described in the `src` directory are also imported below.

In [28]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from ersilia import ErsiliaModel
import sys
sys.path.append('src')
from smiles_processing import standardise_smiles
from inchikey_processing import standardise_inchikey

# Data Preprocessing

The dataset was downloaded from ChEMBL in tsv format. Dataset contains 8715 entries. Thus, it is preprocessed to check for null or irrelevant data and satisfy the required format.

## Reading the raw data

In [11]:
df = pd.read_csv('../data/raw_data.tsv', delimiter='\t')

## Exploring the raw data

In [12]:
df.head()

ChEMBL ID Name           Synonyms            Type  Max Phase  \
0  CHEMBL1964670  NaN                NaN  Small molecule        NaN   
1  CHEMBL3120732  NaN                NaN  Small molecule        NaN   
2    CHEMBL44100  NaN                NaN  Small molecule        NaN   
3   CHEMBL111759  NaN  N-Prenyl Agmatine  Small molecule        NaN   
4  CHEMBL1968584  NaN                NaN  Small molecule        NaN   

   Molecular Weight  Targets  Bioactivities  AlogP  Polar Surface Area  ...  \
0            132.16     53.0           53.0   0.32               46.53  ...   
1            198.27      2.0            3.0   0.57               58.20  ...   
2            179.18      5.0            5.0   0.72               80.36  ...   
3            198.31      3.0           11.0   0.81               73.93  ...   
4            140.14     56.0           56.0   0.75               54.37  ...   

   Heavy Atoms  HBA (Lipinski)  HBD (Lipinski)  #RO5 Violations (Lipinski)  \
0            9               3               1                           0   
1           14               4               2                           0   
2           13               5               4                           0   
3           14               4               5                           0   
4           10               3               1                           0   

  Molecular Weight (Monoisotopic)  Np Likeness Score  Molecular Species  \
0                        132.0786               0.45            NEUTRAL   
1                        198.1368              -0.39            NEUTRAL   
2                        179.0695              -0.51               BASE   
3                        198.1844               1.25               BASE   
4                        140.0473               0.47               ACID   

   Molecular Formula                    Smiles                    Inchi Key  
0            C6H12O3              CCOC(=O)CCCO  AYPJVXQBVHCUCJ-UHFFFAOYSA-N  
1         C10H18N2O2  CC(=O)NC1CCC(NC(C)=O)CC1  TVILGUBKPIUIFC-UHFFFAOYSA-N  
2           C8H9N3O2    N=C(N)Nc1ccc2c(c1)OCO2  OZINHCFSCQEUPW-UHFFFAOYSA-N  
3           C10H22N4      CC(C)=CCNC(=N)NCCCCN  VACJHYSOFQSYAC-UHFFFAOYSA-N  
4             C7H8O3       CC(O)=C1C(=O)CCC1=O  AJIYRXFNAHUXKP-UHFFFAOYSA-N  

[5 rows x 33 columns]

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8715 entries, 0 to 8714
Data columns (total 33 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   ChEMBL ID                        8715 non-null   object 
 1   Name                             670 non-null    object 
 2   Synonyms                         1168 non-null   object 
 3   Type                             8715 non-null   object 
 4   Max Phase                        243 non-null    float64
 5   Molecular Weight                 8715 non-null   float64
 6   Targets                          7689 non-null   float64
 7   Bioactivities                    7689 non-null   float64
 8   AlogP                            8715 non-null   float64
 9   Polar Surface Area               8715 non-null   float64
 10  HBA                              8715 non-null   int64  
 11  HBD                              8715 non-null   int64  
 12  #RO5 Violations     

We only require the `Smiles` and `Inchi Key` columns for further analysis. So we remove all the other columns except these.

In [14]:
selected_columns = ['Smiles', 'Inchi Key']
df = df.loc[:, selected_columns]

In [15]:
df.head()

Molecular Formula                    Smiles                    Inchi Key
0           C6H12O3              CCOC(=O)CCCO  AYPJVXQBVHCUCJ-UHFFFAOYSA-N
1        C10H18N2O2  CC(=O)NC1CCC(NC(C)=O)CC1  TVILGUBKPIUIFC-UHFFFAOYSA-N
2          C8H9N3O2    N=C(N)Nc1ccc2c(c1)OCO2  OZINHCFSCQEUPW-UHFFFAOYSA-N
3          C10H22N4      CC(C)=CCNC(=N)NCCCCN  VACJHYSOFQSYAC-UHFFFAOYSA-N
4            C7H8O3       CC(O)=C1C(=O)CCC1=O  AJIYRXFNAHUXKP-UHFFFAOYSA-N

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8715 entries, 0 to 8714
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Molecular Formula  8715 non-null   object
 1   Smiles             8715 non-null   object
 2   Inchi Key          8715 non-null   object
dtypes: object(3)
memory usage: 204.4+ KB


# Standardising Smiles and Inchi Key

We standardise both `Smiles` and `InChiKeys` of the molecules using the functions described in the `src` directory

In [18]:
smiles_list = df['Smiles'].tolist()
standardised_smiles_list = standardise_smiles(smiles_list)
df['standardised_smiles'] = standardised_smiles_list
df

[22:59:22] Can't kekulize mol.  Unkekulized atoms: 3 5
[22:59:22] Can't kekulize mol.  Unkekulized atoms: 3 9
[22:59:22] Can't kekulize mol.  Unkekulized atoms: 3 9
[22:59:22] Can't kekulize mol.  Unkekulized atoms: 3 5
[22:59:22] Can't kekulize mol.  Unkekulized atoms: 3 7
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 3 6
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 3 7
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 3 7
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 2 5 9 10 11 12
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 2 5 10 11 12 13
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 3 6
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 3 5
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 3 7
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 3 7
[22:59:23] Can't kekulize mol.  Unkekulized atoms: 3 7
[22:59:24] Can't kekulize mol.  Unkekulized atoms: 3 5
[22:59:25] Can't kekulize mol.  Unkekulized atoms: 3 4 5 6 8 9
[22:59:25] Can't kekulize mol.  Un

Molecular Formula                         Smiles  \
0              C6H12O3                   CCOC(=O)CCCO   
1           C10H18N2O2       CC(=O)NC1CCC(NC(C)=O)CC1   
2             C8H9N3O2         N=C(N)Nc1ccc2c(c1)OCO2   
3             C10H22N4           CC(C)=CCNC(=N)NCCCCN   
4               C7H8O3            CC(O)=C1C(=O)CCC1=O   
...                ...                            ...   
8710         C5H11NO2S         CSCC[C@H](N)[11C](=O)O   
8711        C4H8NNaOS2           OCCCNC(=S)[S-].[Na+]   
8712            C6H8O3        [3H]C1(O)C=C(C(=O)O)CC1   
8713          C8H11NO3  [3H][C@](O)(CN)c1ccc(O)c(O)c1   
8714        C9H13N2O2+      CN(C)C(=O)Oc1ccc[n+](C)c1   

                        Inchi Key            standardised_smiles  
0     AYPJVXQBVHCUCJ-UHFFFAOYSA-N                   CCOC(=O)CCCO  
1     TVILGUBKPIUIFC-UHFFFAOYSA-N       CC(=O)NC1CCC(NC(C)=O)CC1  
2     OZINHCFSCQEUPW-UHFFFAOYSA-N         N=C(N)Nc1ccc2c(c1)OCO2  
3     VACJHYSOFQSYAC-UHFFFAOYSA-N           CC(C)=CCNC(=N)NCCCCN  
4     AJIYRXFNAHUXKP-UHFFFAOYSA-N            CC(=O)C1C(=O)CCC1=O  
...                           ...                            ...  
8710  FFEARJCKVFRZRR-YMNXMHMNSA-N         CSCC[C@H](N)[11C](=O)O  
8711  DKXSDDDDKKDNIP-UHFFFAOYSA-M                    OCCCNC(=S)S  
8712  VKJGQKRRULDFCI-XHHURNKPSA-N        [3H]C1(O)C=C(C(=O)O)CC1  
8713  SFLSHLFXELFNJZ-CMIMLBRMSA-N  [3H][C@](O)(CN)c1ccc(O)c(O)c1  
8714  RVOLLAQWKVFTGE-UHFFFAOYSA-N      CN(C)C(=O)Oc1ccc[n+](C)c1  

[8715 rows x 4 columns]

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8715 entries, 0 to 8714
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Molecular Formula    8715 non-null   object
 1   Smiles               8715 non-null   object
 2   Inchi Key            8715 non-null   object
 3   standardised_smiles  8697 non-null   object
dtypes: object(4)
memory usage: 272.5+ KB


Since few molecules could not be kekulized, we remove those rows from our dataset as they contain `NULL` entries in the `standardised_smiles` column.

In [20]:
df.dropna(inplace=True)
df.reset_index(drop=True, inplace=True)
df

Molecular Formula                         Smiles  \
0              C6H12O3                   CCOC(=O)CCCO   
1           C10H18N2O2       CC(=O)NC1CCC(NC(C)=O)CC1   
2             C8H9N3O2         N=C(N)Nc1ccc2c(c1)OCO2   
3             C10H22N4           CC(C)=CCNC(=N)NCCCCN   
4               C7H8O3            CC(O)=C1C(=O)CCC1=O   
...                ...                            ...   
8692         C5H11NO2S         CSCC[C@H](N)[11C](=O)O   
8693        C4H8NNaOS2           OCCCNC(=S)[S-].[Na+]   
8694            C6H8O3        [3H]C1(O)C=C(C(=O)O)CC1   
8695          C8H11NO3  [3H][C@](O)(CN)c1ccc(O)c(O)c1   
8696        C9H13N2O2+      CN(C)C(=O)Oc1ccc[n+](C)c1   

                        Inchi Key            standardised_smiles  
0     AYPJVXQBVHCUCJ-UHFFFAOYSA-N                   CCOC(=O)CCCO  
1     TVILGUBKPIUIFC-UHFFFAOYSA-N       CC(=O)NC1CCC(NC(C)=O)CC1  
2     OZINHCFSCQEUPW-UHFFFAOYSA-N         N=C(N)Nc1ccc2c(c1)OCO2  
3     VACJHYSOFQSYAC-UHFFFAOYSA-N           CC(C)=CCNC(=N)NCCCCN  
4     AJIYRXFNAHUXKP-UHFFFAOYSA-N            CC(=O)C1C(=O)CCC1=O  
...                           ...                            ...  
8692  FFEARJCKVFRZRR-YMNXMHMNSA-N         CSCC[C@H](N)[11C](=O)O  
8693  DKXSDDDDKKDNIP-UHFFFAOYSA-M                    OCCCNC(=S)S  
8694  VKJGQKRRULDFCI-XHHURNKPSA-N        [3H]C1(O)C=C(C(=O)O)CC1  
8695  SFLSHLFXELFNJZ-CMIMLBRMSA-N  [3H][C@](O)(CN)c1ccc(O)c(O)c1  
8696  RVOLLAQWKVFTGE-UHFFFAOYSA-N      CN(C)C(=O)Oc1ccc[n+](C)c1  

[8697 rows x 4 columns]

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8697 entries, 0 to 8696
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Molecular Formula    8697 non-null   object
 1   Smiles               8697 non-null   object
 2   Inchi Key            8697 non-null   object
 3   standardised_smiles  8697 non-null   object
dtypes: object(4)
memory usage: 271.9+ KB


We standardise the `Inchi Key` column also using the function described in the `src` directory.

In [22]:
smiles_list = df['standardised_smiles'].tolist()
standardised_inchikeys_list = standardise_inchikey(smiles_list)
df['standardised_inchikeys'] = standardised_inchikeys_list
df

Molecular Formula                         Smiles  \
0              C6H12O3                   CCOC(=O)CCCO   
1           C10H18N2O2       CC(=O)NC1CCC(NC(C)=O)CC1   
2             C8H9N3O2         N=C(N)Nc1ccc2c(c1)OCO2   
3             C10H22N4           CC(C)=CCNC(=N)NCCCCN   
4               C7H8O3            CC(O)=C1C(=O)CCC1=O   
...                ...                            ...   
8692         C5H11NO2S         CSCC[C@H](N)[11C](=O)O   
8693        C4H8NNaOS2           OCCCNC(=S)[S-].[Na+]   
8694            C6H8O3        [3H]C1(O)C=C(C(=O)O)CC1   
8695          C8H11NO3  [3H][C@](O)(CN)c1ccc(O)c(O)c1   
8696        C9H13N2O2+      CN(C)C(=O)Oc1ccc[n+](C)c1   

                        Inchi Key            standardised_smiles  \
0     AYPJVXQBVHCUCJ-UHFFFAOYSA-N                   CCOC(=O)CCCO   
1     TVILGUBKPIUIFC-UHFFFAOYSA-N       CC(=O)NC1CCC(NC(C)=O)CC1   
2     OZINHCFSCQEUPW-UHFFFAOYSA-N         N=C(N)Nc1ccc2c(c1)OCO2   
3     VACJHYSOFQSYAC-UHFFFAOYSA-N           CC(C)=CCNC(=N)NCCCCN   
4     AJIYRXFNAHUXKP-UHFFFAOYSA-N            CC(=O)C1C(=O)CCC1=O   
...                           ...                            ...   
8692  FFEARJCKVFRZRR-YMNXMHMNSA-N         CSCC[C@H](N)[11C](=O)O   
8693  DKXSDDDDKKDNIP-UHFFFAOYSA-M                    OCCCNC(=S)S   
8694  VKJGQKRRULDFCI-XHHURNKPSA-N        [3H]C1(O)C=C(C(=O)O)CC1   
8695  SFLSHLFXELFNJZ-CMIMLBRMSA-N  [3H][C@](O)(CN)c1ccc(O)c(O)c1   
8696  RVOLLAQWKVFTGE-UHFFFAOYSA-N      CN(C)C(=O)Oc1ccc[n+](C)c1   

           standardised_inchikeys  
0     AYPJVXQBVHCUCJ-UHFFFAOYSA-N  
1     TVILGUBKPIUIFC-UHFFFAOYSA-N  
2     OZINHCFSCQEUPW-UHFFFAOYSA-N  
3     VACJHYSOFQSYAC-UHFFFAOYSA-N  
4     WYBUYWCJZMQESH-UHFFFAOYSA-N  
...                           ...  
8692  FFEARJCKVFRZRR-YMNXMHMNSA-N  
8693  HSXNXTGUYOUXQI-UHFFFAOYSA-N  
8694  VKJGQKRRULDFCI-XHHURNKPSA-N  
8695  SFLSHLFXELFNJZ-CMIMLBRMSA-N  
8696  RVOLLAQWKVFTGE-UHFFFAOYSA-N  

[8697 rows x 5 columns]

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8697 entries, 0 to 8696
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Molecular Formula       8697 non-null   object
 1   Smiles                  8697 non-null   object
 2   Inchi Key               8697 non-null   object
 3   standardised_smiles     8697 non-null   object
 4   standardised_inchikeys  8697 non-null   object
dtypes: object(5)
memory usage: 339.9+ KB


Here, the preprocessing is completed. The dataset now contains 8697 entries. We now create a sample dataset of 1000 entries using the sample function. We also rename the column names to `Smiles` and `Inchi_key` for convenience.

# Creating the Sample of 1000 Entries

In [26]:
input_data = df.sample(n=1000, random_state=42)
input_data = input_data[['standardised_inchikeys', 'standardised_smiles']]
input_data.rename(columns={'standardised_inchikeys': 'Inchi_key', 'standardised_smiles': 'Smiles'}, inplace=True)
input_data.reset_index(drop=True, inplace=True)
input_data.to_csv('../data/input_data.csv', index=False)
input_data.head()

Inchi_key                      Smiles
0    RAEDEZWVLQSYAR-CLFYSBASSA-N  [O-]/[N+](=N\OCCl)N1CCOCC1
1    UTXDPIGHIUIJIS-UHFFFAOYSA-N            CC(CN)c1cnc[nH]1
2    ODJQKYXPKWQWNK-UHFFFAOYSA-N           O=C(O)CCSCCC(=O)O
3    BWUFFNRDKQQEQQ-GDVGLLTNSA-N   CC1SC(=O)[C@@H]2CCCN2C1=O
4    RSXWJXPKLRYMHW-UHFFFAOYSA-N   Cc1nc([N+](=O)[O-])cn1CCO
..                           ...                         ...
995  DCUKVUVLSDGDEZ-UHFFFAOYSA-N              CCN(N=O)C(C)OC
996  NDVNHAJMXQWLSL-UHFFFAOYSA-N          CCNc1ncnc2c1nnn2CC
997  ALYNCZNDIQEVRV-UHFFFAOYSA-N           Nc1ccc(C(=O)O)cc1
998  COAVYXJTHCOFKX-UHFFFAOYSA-N       CC(C)C(O)(CCO)CC(=O)O
999  JUZLMRLZYMOOMY-XQRVVYSFSA-N    CN(C)/C=C1/C=C(C=O)OC1=O

[1000 rows x 2 columns]

In [62]:
input_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Inchi_key  1000 non-null   object
 1   Smiles     1000 non-null   object
dtypes: object(2)
memory usage: 15.8+ KB


# Running the Model on the Input Data

The model is run on the terminal on the input data using the following commands:
```
ersilia -v fetch eos30gr
ersilia serve eos30gr
ersilia -v api run -i input_data.csv -o output_data.csv
```

The output after running the model is saved in `output_data.csv` file which is present the `data` directory.

# Model Bias Evaluation

## Exploring the output data

In [60]:
predictions_df = pd.read_csv('../data/output_data.csv')
predictions_df.head()

key                       input  activity10
0  RAEDEZWVLQSYAR-CLFYSBASSA-N  [O-]/[N+](=N\OCCl)N1CCOCC1    0.342822
1  UTXDPIGHIUIJIS-UHFFFAOYSA-N            CC(CN)c1cnc[nH]1    0.183168
2  ODJQKYXPKWQWNK-UHFFFAOYSA-N           O=C(O)CCSCCC(=O)O    0.034653
3  BWUFFNRDKQQEQQ-GDVGLLTNSA-N   CC1SC(=O)[C@@H]2CCCN2C1=O    0.224422
4  RSXWJXPKLRYMHW-UHFFFAOYSA-N   Cc1nc([N+](=O)[O-])cn1CCO    0.136492

In [61]:
predictions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   key         1000 non-null   object 
 1   input       1000 non-null   object 
 2   activity10  1000 non-null   float64
dtypes: float64(1), object(2)
memory usage: 23.6+ KB


The output data contains the `key` - which is the `InChiKey`, the `input` - which is the `Smiles` string and the `activity10` column which contains the `probability of hERG blockade`. 

## Visualizing the predictions